### PR TITLE
Add SIA50 support

### LINKED ISSUE #334: SIA50D robot model support
Hi,
Thank you for maintaining this high-quality package!

I was looking into an application using a [SIA50D](https://www.motoman.com/en-us/products/robots/industrial/assembly-handling/sia-series/sia50d). As part of this I am looking for the robot description but it is not yet included in this repository.

Are there plans to add support for this model (even if only description/MoveIt config) to this package?

I am happy to contribute time and effort to adding the robot model but do not have access to meshes/further specifications. If someone from Yaskawa/Motoman follows this repository, or if you have access within ROS-Industrial, I am happy to help getting it added to this package if of interest.

Thanks,
Wolfgang

### PR DESCRIPTION
**Pull request by wxmerkt**
This is a WIP PR to add support for the SIA50D. This PR is based on the support for the SIA20D and the mesh file processing for the GP88. Thanks to @EricMarcil for providing the meshes, model, and instructions.

The automatically created collision mesh/convex hull for the 6th link is currently not great (used the same script as for GP88).

I am planning on adding approximate CoG, inertia, and mass information from CAD assuming uniform mass distribution in each link (as we are planning to use the model for dynamic trajectory optimisation and the correct values are not available for open source release). Please let me know if this is OK or if you prefer to omit the inertial tags in this case.

Looking forward to your comments and valuable feedback, and thanks for all the support,
Wolfgang

Please note: This has _not_ been tested with a real robot.

Resolves #334

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Thanks for the PR.

High-level comment to start with: I'd like to remove the `d` suffix on the robot variant identifier (ie: `sia50d`).

Afaik this encodes for the controller (a DX100 in your case most likely), but afaik, the robot would not be different when used with an FS100 (I can't find it with an FS100 though).

@EricMarcil / @ted-miller: could you confirm?

There are already too many support packages which differentiate on the controller (ie: `motoman_sia5d_support` and `motoman_sia5f_support`), but most of the times it's the same robot.

See also #65.

Additionally: if you'd really like to prefix your commit msgs with the pkg name, please drop the `motoman_` prefix. All commits are pushed to the `motoman` repository, so everything here is a `motoman_` package. We like to avoid repeating ourselves, so no need for that `motoman_` there again. As a suggestion: `sia50_support: <your commit comment>`.

File: motoman_sia50d_support/test/launch_test.xml
Comment (gavanderhoorn) [inline motoman_sia50d_support/test/launch_test.xml]: CI is likely failing because of this line (and the same one below for the `visualize` launch file).

You're passing the `controller` arg, but it's not actually something you can set: it already has a `value` (namely: `dx100`).

**Comment by wxmerkt** [comment]
Thanks for the review and feedback. I added a fix for the CI and will monitor it. From the website, I believe the SIA50D exists only with the DX100 controller - there is no F variant listed: https://www.motoman.com/en-us/products/robots/industrial/assembly/sia
I'll wait for feedback from Motoman and will rename the package to drop the "d" postfix then accordingly.

As for adding approximate inertial properties, is that okay or would you prefer to omit them?

**Comment by gavanderhoorn** [comment]
> From the website, I believe the SIA50D exists only with the DX100 controller

That's not really the point: from other robots we know that the last letter encodes the controller. With other support pkgs we also don't add that to the variant identifier (here in this repository, nor anywhere else in ROS-Industrial).

Tbh we never should've done that with the SIA5 or the SDA series either.

> As for adding approximate inertial properties, is that okay or would you prefer to omit them?

I would personally remove them, but it depends a bit on how inaccurate they are: if they lead to OK behaviour in Gazebo, we could keep them, but we would have to explicitly state they are approximations.

**Comment by wxmerkt** [comment]
Thanks for the feedback. I've renamed the package.

I've committed the approximate inertial parameters in https://github.com/ros-industrial/motoman/pull/337/commits/79e3d84e3000e9919151f4364a336f0ed831e1e3. These values have been obtained by using the full mass of the manipulator according to spec sheet (650kg), and assuming a uniform mass distribution across all links, and then derived from CAD. If @EricMarcil could please comment if they are approximately OK I would leave them in the PR; otherwise I will revert/drop this commit from the PR. (We have not validated them in Gazebo but for motion planning using full rigid-body dynamics. If there is a Gazebo package for Motoman, I am happy to adapt/test it with it).

**Comment by EricMarcil** [comment]
@gavanderhoorn Though it may no longer be relevant, to answer your question, the SDA50 is only available with the DX100 since the power requirement are to big for the FS100 which is a compact controller for smaller robots.

**Comment by gavanderhoorn** [comment]
> @gavanderhoorn Though it may no longer be relevant, to answer your question, the SDA50 is only available with the DX100 since the power requirement are to big for the FS100 which is a compact controller for smaller robots.

Thanks @EricMarcil.

It's more the general naming in this case, not so much whether this specific robot would be available with a specific controller.

re: power requirements: yes, I'm somewhat jealous of @wxmerkt having access to an SIA50.

**Comment by EricMarcil** [comment]
@wxmerkt First, thank you for you contribution.
Concerning the mass and other properties, I haven't gone into details but I can quickly tell you that the URB mass is over evaluated by more than 50%.  
Unfortunately, note that I will not comment any further on this topic of estimating these values, since this information are considered confidential.  If you need further details for your application then I would invite you to contact your local Yaskawa representative and establish a NDA with them.  Once that is establish, they can (or I can) directly assist you.

File: motoman_sia50_support/launch/robot_interface_streaming_sia50d.launch
Comment (gavanderhoorn) [inline motoman_sia50_support/launch/robot_interface_streaming_sia50d.launch]: Could you make this an `arg` with a `default` instead?

It'd better align with the other `.launch` files.

We could also opt for removing the `arg` completely, but that would mean we'd have to change these files again if/when Yaskawa decides that there is going to be a SIA50 with a YRC1000 (or something like that).

File: motoman_sia50_support/launch/robot_state_visualize_sia50d.launch
Comment (gavanderhoorn) [inline motoman_sia50_support/launch/robot_state_visualize_sia50d.launch]: Same here: please `default` instead of `value`.

File: motoman_sia50_support/launch/test_sia50d.launch
Comment (gavanderhoorn) [inline motoman_sia50_support/launch/test_sia50d.launch]: We'll deal with the `joint_state_publisher_gui` issue in a later PR, for all support packages.

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]:
```suggestion
    <p>ROS-Industrial support for the Motoman SIA50 (and variants).</p>
```

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]: Please mention document title and number/revision/ID instead of linking to a website.

The link will go stale.

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]: I appreciate you adding this, but let's make it a regular paragraph in the manifest: revert the all-caps and remove the `strong`.

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]: Are you sure you are going to be maintaining this?

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]: Please remove.

URDFs/`.xacro`s are so 'minimal' that we don't retain authorship of originals in derivatives, especially not as you would've ended up with the same file if you'd use the templates in the support package generator.

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]: Remove.

File: motoman_sia50_support/package.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/package.xml]:
```suggestion
  <maintainer email="g.a.vanderhoorn@tudelft.nl">G.A. vd. Hoorn (TU Delft Robotics Institute)</maintainer>
```

File: motoman_sia50_support/test/launch_test.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/test/launch_test.xml]: Remove

File: motoman_sia50_support/test/launch_test.xml
Comment (gavanderhoorn) [inline motoman_sia50_support/test/launch_test.xml]: Unless you make `controller` a real `arg` again in the `.launch` files in which it is used, this can be removed.

**Comment by wxmerkt** [comment]
> re: power requirements: yes, I'm somewhat jealous of @wxmerkt having access to an SIA50.

Alas, we don't have access to one (thus this is all simulation only). We need the model to replicate an experimental setup for comparison that uses a SIA50. The 2m+ workspace and 7-DoF class is quite a narrow subset of the overall market.

> Concerning the mass and other properties, I haven't gone into details but I can quickly tell you that the URB mass is over evaluated by more than 50%.
Thanks for the feedback, @EricMarcil. I have removed the inertial properties from the pull request. 

Looking forward to your other feedback on the pull request and happy to amend as required.

File: motoman_sia50_support/urdf/sia50d.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d.xacro]: Please rename this file to remove the `d` suffix, as there is nothing DX100 specific in this file.

Make sure to update the `name` attribute of `robot` and the xacro macro name as well.

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Rename this file as well.

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Remove the `d` suffix.

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Please also remove the `d` suffix from the `meshes` sub dir.

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Please use the `radians(..)` conversion factor and specify limits in degrees (so they can be easily compared to the specs).

Additionally: `18.5` Nm for the first joint? This seems rather low.

If you want to include an `effort` limit here, you should probably list the other number (ie: maximum acc/dec torque).

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Is there a specific reason you have the Z translation in 6 significant digits here?

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Ideally `joint`s have translations in a single dimension only. None of the other `joint`s in your macro have this double translation.

Is it required?

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Please remove these comments.

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (gavanderhoorn) [inline motoman_sia50_support/urdf/sia50d_macro.xacro]: Could you clarify why this is commented?

We should not merge any more macros without the `base` and `tool0` frames defined.

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
Thanks again for the PR.

I've left some comments.

Overall: could you please check whether all files use spaces for indents, not tabs? Please use 2-space indents per level in all XML-based files.

File: motoman_sia50_support/package.xml
Comment (wxmerkt) [reply motoman_sia50_support/package.xml]: I used the [image](https://www.motoman.com/getmedia/3fb46998-8d43-4afc-b6f0-b88cfe21e33b/sia50d_specs.jpg.aspx) on the "Specifications" tab. As thus, I don't have the document title/number/revision/ID. I also used a `.csv` sent by @EricMarcil for limits that did not come with a revision or version.

Comment (gavanderhoorn) [reply motoman_sia50_support/package.xml]: All the documentation of Yaskawa is available free-of-charge on their website. It should not be difficult to locate the specifications of the SIA50 and mention the document ID here.

File: motoman_sia50_support/urdf/sia50d_macro.xacro
Comment (wxmerkt) [reply motoman_sia50_support/urdf/sia50d_macro.xacro]: It's due to this:

image

image

Comment (gavanderhoorn) [reply motoman_sia50_support/urdf/sia50d_macro.xacro]: This is solvable by placing `link_s` in the same ~~place~~ *plane* as `link_l`. That's beneficial, as that would also be the location of the `base` frame.

**Comment by wxmerkt** [comment]
Thank you for the detailed review and feedback @gavanderhoorn. I've addressed the comments

**Comment by EricMarcil** [comment]
Thanks for the PR.  It is greatly appreciated.
I've review it and everything looks good on the robot side.
However there is a "Tools" folder that I think is specific to your application that should be removed from the PR.

**Comment by wxmerkt** [comment]
Thank you for the review and test on hardware. I am happy to remove the `tools` folder - it mirrors the processing scripts from `GP88`. Would you like me to remove it (it might be used to adapt/reproduce the meshes from the originals in the future)?

**Comment by EricMarcil** [comment]
Note, I did not actually test it on hardware.  I just compared the values with my simulator (MotoSim).
As for the Tools folder, I'm wasn't aware of those processing scripts.  I just assumed it was related to your tooling or something.
@gavanderhoorn : When available, should the Tools folder be included in the support package?

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
Overall this seems OK, but the collision meshes are a bit strange:

rviz_screenshot_2020_07_28-20_30_47

Did you have a particular reason to create them like this @wxmerkt?

If not, I'd recommend to create convex hulls using Meshlab from the `visual` meshes.

Additionally, there are two meshes with missing faces.

`link_l`:

rviz_screenshot_2020_07_28-20_33_59

rviz_screenshot_2020_07_28-20_34_15

and `link_b`:

rviz_screenshot_2020_07_28-20_36_15

Those meshes will have to be fixed before we can merge.

**Comment by gavanderhoorn** [comment]
@EricMarcil wrote:
> As for the Tools folder, I'm wasn't aware of those processing scripts. I just assumed it was related to your tooling or something.
> @gavanderhoorn : When available, should the Tools folder be included in the support package?

We can keep them: they (partially) document how the meshes were generated/converted from the original ones. Assuming @wxmerkt actually used them, of course.

**Comment by EricMarcil** [comment]
@wxmerkt : I've managed to find another model that wasn't missing faces in the stl mesh.  I've regenerated the meshes, so if you want to substitute them to yours and confirm that everything is displaying properly.
[SIA50D STL.zip](https://github.com/ros-industrial/motoman/files/5091949/SIA50D.STL.zip)

**Comment by gavanderhoorn** [comment]
@wxmerkt: friendly ping.

It's ok if you can't spend any more time on this. If you can let us now, I'll take over and fix the meshes.

**Comment by wxmerkt** [comment]
Hi @gavanderhoorn,
Sorry for the silence on this - I can't spend any more time on this right now - thank you for the offer to take over and complete the PR. Regarding the earlier questions: I used the scripts in the processing folder. They use Meshlab under the hood as with the folder I amended them from (GP88). The artefacts in the collision meshes thus must be from needing to amend the settings (I used the very latest Meshlab).

Many thanks for all your input, time and help with this PR

**Comment by gavanderhoorn** [comment]
I've fixed all outstanding issues with the meshes (thanks @EricMarcil for the new set) and added a few more additional commits.

Once Travis gives us the green checkmark, I'll squash-merge.

**Comment by gavanderhoorn** [comment]
And in the end I've removed the `tools` folder, as the meshes we're finally using were not produced using those scripts.

**Review by gavanderhoorn — APPROVED** [review]
With all outstanding issues fixed I believe we can merge this.

**Comment by gavanderhoorn** [comment]
@wxmerkt: thanks for the (initial) PR and thanks for iterating.

@EricMarcil: thanks for the meshes and overall checks and guidance.

**Comment by gavanderhoorn** [comment]
@wxmerkt: due to the use of a squash-merge, it doesn't show, but you could remove your `wxm-add-motoman-sia50d-support` branch.

**Pull request merged**
